Introduce O(n^1.5) BigDecimal parser implementation

ferenc-csaky opened the pull request:
Covers the `BigDecimal` part of #577. The `double`/`float` implementations would require a much bigger refactor, since the example implementation follows a very different approach, but IMO it's worth to update only the `BigDecimal` without the others as well.

cowtowncoder commented:
@ferenc-csaky Sounds good -- thank you very much for the contribution! -- and it looks like I have CLA from you already.

Just one request: could this be rebased against 2.13? `master` is for Jackson 3.0 and I think this improvement would be great for 2.13 -- I can handle merging to `master`.

ferenc-csaky commented:
Changed the base branch to `2.13`. My original plan was this as well, my dev. branch also originated from `2.13`, but it seems I missed it in the PR. Corrected now, thanks!

cowtowncoder commented:
Just realized that none of the tests verifies new code path(s), I think. Will repurpose some from original... should have caught this earlier,.

cowtowncoder commented:
@ferenc-csaky  Hmmh. I think there may be a bug somewhere, some offset not being carried or used for calculations?
I added a test in 2.13 from the original changes and it looks like offset used for splits is `0` for some reason.
As far as I can see it gets passed correctly to `BigDecimalParser` (first suspected it might not but seems to).

One thing that is tricky, too, is how to propagate underlying secondary fails, if those are possible (not sure if they are after fixing this).

cowtowncoder commented:
Looks like `int numIdx = 0` in `parseBigDecimal(int splitLen)` is wrong: that should get set to `off`, somewhere. But simple attempts did not produce expected results.

ferenc-csaky commented:
Oh yes, you are totally right, I didn't follow the source impl. exactly and I messed it up. Sorry about it. I just created a PR with the fix. Now the test you added runs correctly.

cowtowncoder commented:
@ferenc-csaky np. Just glad this was caught :)
